### PR TITLE
Rule S6422: Calls to "async" methods should not be blocking in Azure Functions

### LINKED ISSUE #5598: Rule S6422: Calls to "async" methods should not be blocking in Azure Functions
MMF: https://jira.sonarsource.com/browse/MMF-2699

RSPEC PR to be merged after implementing this ticket: https://github.com/SonarSource/rspec/pull/957

* See implementation for Rule S4462 ([CallToAsyncMethodShouldNotBeBlocking](https://github.com/SonarSource/sonar-dotnet/blob/8.38.0.46746/analyzers/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs#L34)).
* This ticket should also modify S4462 to not raise on Azure Functions - modification to the S4462 RSPEC to add as exception https://github.com/SonarSource/rspec/pull/967
* Probably the same implementation in CallToAsyncMethodShouldNotBeBlocking can be used for both, and raise S6422 or S4462 depending on the context.

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5598

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs [hunk 1]
```diff
@@ -107,45 +101,35 @@ private static void ReportOnViolation(SyntaxNodeAnalysisContext context)
...
                 }
-
                 if (memberAccessNameName == ResultName && IsAwaited(context, simpleMemberAccess))
                 {
                     return;  // No need to report an issue on a waited object
                 }
             }
-
-            context.ReportIssue(Diagnostic.Create(Rule, simpleMemberAccess.GetLocation(), MemberNameToMessageArguments[memberAccessNameName]));
+            context.ReportIssue(Diagnostic.Create(context.IsAzureFunction() ? RuleS6422 : RuleS4462, simpleMemberAccess.GetLocation(), MemberNameToMessageArguments[memberAccessNameName]));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs:122]: This is only functional change. The rest is only a mechanical refactoring.

File: analyzers/rspec/cs/S6422_c#.html [hunk 2]
```diff
@@ -0,0 +1,73 @@
+<p>Making blocking calls to <code>async</code> methods transforms something that was intended to be asynchronous into a synchronous block. Doing so
+inside an Azure Function can lead to thread exhaustion.</p>
```
Comment (martin-strecker-sonarsource) [inline analyzers/rspec/cs/S6422_c#.html:2]: >lead to thread exhaustion

Is this backed by some research? Azure Function may not be affected by that or the mitigation strategies are different in that environment.

File: analyzers/rspec/cs/S6422_c#.html [hunk 3]
```diff
@@ -0,0 +1,73 @@
+<p>Making blocking calls to <code>async</code> methods transforms something that was intended to be asynchronous into a synchronous block. Doing so
```
Comment (martin-strecker-sonarsource) [inline analyzers/rspec/cs/S6422_c#.html:1]: "something" and "synchronous block" sound weird. Maybe like so?
```suggestion
<p>Making blocking calls to <code>async</code> methods transforms code that was intended to be asynchronous into a blocking operation. Doing so
```

File: analyzers/src/SonarAnalyzer.Common/Extensions/SyntaxNodeAnalysisContextExtensions.cs [hunk 4]
```diff
@@ -39,6 +39,9 @@ internal static class SyntaxNodeAnalysisContextExtensions
         internal static bool IsRedundantPositionalRecordContext(this SyntaxNodeAnalysisContext context) =>
             context.ContainingSymbol.Kind == SymbolKind.Method;
 
+        public static bool IsAzureFunction(this SyntaxNodeAnalysisContext context) =>
+            context.AzureFunctionMethod() is not null;
+
         public static IMethodSymbol AzureFunctionMethod(this SyntaxNodeAnalysisContext context) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/SyntaxNodeAnalysisContextExtensions.cs:45]: NIT: Consider swapping the order of IsAzureFunction and AzureFunctionMethod

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs [hunk 5]
```diff
@@ -173,16 +159,13 @@ private static bool IsAwaited(SyntaxNodeAnalysisContext context, MemberAccessExp
...
                 _ => false
             };
 
         private static bool IsChainedAfterThreadPoolCall(MemberAccessExpressionSyntax memberAccess, SemanticModel semanticModel) =>
-            memberAccess.Expression.DescendantNodes()
-                                   .OfType<MemberAccessExpressionSyntax>()
-                                   .Any(subMemberAccess => IsNamedSymbolOfExpectedType(subMemberAccess, semanticModel, TaskThreadPoolCalls));
+            memberAccess.Expression.DescendantNodes().OfType<MemberAccessExpressionSyntax>().Any(x => IsNamedSymbolOfExpectedType(x, semanticModel, TaskThreadPoolCalls));
 
         private static bool IsNamedSymbolOfExpectedType(MemberAccessExpressionSyntax memberAccess, SemanticModel semanticModel, Dictionary<string, KnownType> expectedTypes) =>
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs:170]: Parameter order: semanticModel should be first.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs [hunk 6]
```diff
@@ -33,60 +33,54 @@ namespace SonarAnalyzer.Rules.CSharp
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     public sealed class CallToAsyncMethodShouldNotBeBlocking : SonarDiagnosticAnalyzer
     {
-        private const string DiagnosticId = "S4462";
         private const string MessageFormat = "Replace this use of '{0}' with '{1}'.";
         private const string ResultName = "Result";
         private const string ContinueWithName = "ContinueWith";
         private const string SleepName = "Sleep";
+        private const string AzureFunctionSuffix = @" Calls to ""async"" methods should not be blocking in Azure Functions.";
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/CallToAsyncMethodShouldNotBeBlocking.cs:40]: This is not precise and does not fit in cases like "WaitAny". You don't await a method result, but a Task (or even more precise "an awaitable object"). Maybe just 
```suggestion
        private const string AzureFunctionSuffix = @" Don't perform blocking operations in Azure Functions.";
```

**Review by martin-strecker-sonarsource — APPROVED** [review]
No functional changes are required, but I have some comments about the wording.

**Comment by martin-strecker-sonarsource** [comment]
LGTM. Assigned you because of some NITs and a merge conflict.

File: analyzers/rspec/cs/S6422_c#.html (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/rspec/cs/S6422_c#.html:2]: I have no idea, let's ask Andrei on slack

File: analyzers/src/SonarAnalyzer.Common/Extensions/SyntaxNodeAnalysisContextExtensions.cs (hunk 4 above)
Comment (MaStr11) [reply analyzers/src/SonarAnalyzer.Common/Extensions/SyntaxNodeAnalysisContextExtensions.cs:45]: Both are public and the former one uses the later. It's easier to read (and F# actually enforces this) when both methods are kept in "depends on order".

**Pull request merged**